spharpy/spherical.py

```diff
@@ -1,104 +1,585 @@
+        self._check_properties(compute='basis_gradient')
+        return self._basis_gradient_theta
+
+    @property
+    def basis_gradient_phi(self):
+        if self._basis_gradient_phi is None:
+            self._compute_basis_gradient()
+        self._check_properties(compute='basis_gradient')
+        return self._basis_gradient_phi
+
+    # @lru_cache(maxsize=128)
+    def _compute_basis(self):
+        """
+        Compute the basis matrix for the SphericalHarmonics class.
+        """
+        logger.info("Computing basis matrix for n_max=%d", self.n_max)
+        if self.basis_type == "complex":
+            self.basis = spherical_harmonic_basis(
+                self.n_max, self.coords,
+                self.normalization, self.channel_convention
+            )
+        elif self.basis_type == "real":
+            self.basis = spherical_harmonic_basis_real(
+                self.n_max, self.coords,
+                self.normalization, self.channel_convention
+            )
+        else:
+            raise ValueError(
+                "Invalid signal type, should be either 'complex' or 'real'"
+            )
+
+    def compute_basis(self):
+        try:
+            return self._compute_basis()
+        except Exception as e:
+            raise ValueError("Error computing basis:", e) from e
+
+    # @lru_cache(maxsize=128)
+    def _compute_basis_gradient(self):
+        logger.info("Computing basis gradient for n_max=%d", self.n_max)
+        if any(
+            (self.normalization in ["maxN", "sn3d"],
+             self.channel_convention == "fuma")
+        ):
+            raise ValueError(
+                "Gradient computation not supported for MaxN, "
+                "SN3D normalization or FuMa channel ordering"
+            )
+        else:
+            if self.basis_type == "complex":
+                grad_theta, grad_phi = spherical_harmonic_basis_gradient(
+                    self.n_max, self.coords
+                )
+                self.basis_gradient_theta, self.basis_gradient_phi = (
+                    grad_theta,
+                    grad_phi,
+                )
+            elif self.basis_type == "real":
+                grad_theta, grad_phi = spherical_harmonic_basis_gradient_real(
+                    self.n_max, self.coords
+                )
+                self.basis_gradient_theta, self.basis_gradient_phi = (
+                    grad_theta,
+                    grad_phi,
+                )
+            else:
+                raise ValueError(
+                    "Invalid signal type, should be either 'complex' or 'real'"
+                )
+
+    def compute_basis_gradient(self):
+        try:
+            return self._compute_basis_gradient()
+        except Exception as e:
+            raise ValueError("Error computing basis gradient:", e) from e
+
+    @property
+    def basis_inv(self):
+        if self._basis_inv is None:
+            self.compute_inverse()
```

Comment:
```suggestion
            self._compute_inverse()
```
I think we can get rid of the public compute inverse function and use the private one here.